text/0000-layout-packed-aligned.md

```diff
@@ -0,0 +1,182 @@
+[guide-level-explanation]: #guide-level-explanation
+
+## `#[repr(C)]`
+When `align` and `packed` attributes exist on the same type, or when `packed` structs transitively contains `align` types,
+the resulting layout matches the target toolchain ABI.
+
+For example, given:
+```c
+#[repr(C, align(4))]
+struct Foo(u8);
+#[repr(C, packed(1))]
+struct Bar(Foo);
+```
+`align_of::<Bar>()` would be 4 for `*-pc-windows-msvc` and 1 for everything else.
+
+
+## `#[repr(system)]`
+When `align` and `packed` attributes exist on the same type, or when `packed` structs transitively contains `align` types,
+the resulting layout matches the target OS ABI.
+
+For example, given:
+```c
+#[repr(system, align(4))]
+struct Foo(u8);
+#[repr(system, packed(1))]
+struct Bar(Foo);
+```
+`align_of::<Bar>()` would be 4 for `*-pc-windows-msvc` and `*-pc-windows-gnu`. It would be 1 for everything else.
+
+## `#[repr(Rust)]`
+When `align(N)` and `packed(M)` attributes exist on the same type, or when `packed` structs contain `aligned` fields,
+the type will have their base alignment increased to `N`, while the struct fields will be laid out as if their
+alignments were decreased to `M`. However, in general Rust is free to reorder
+these fields for optimization purposes, and the only guarantee is that the fields will maintain a minimum alignment of `M`.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+In the following paragraphs, "Decreasing M to N" means:
+```
+if M > N {
+    M = n
+}
+```
+
+"Increasing M to N" means:
+```
+if M < N {
+    M = N
+}
+```
+
+
+`#[repr(align(N))]` increases the base alignment of a type to be N.
+
+`#[repr(packed(M))]` decreases the alignment of the struct fields to be M. Because the base alignment of the type
+is defined as the maximum of the alignment for any fields, this also has the indirect result of decreasing the base
+alignment of the type to be M.
+
+When the align and packed modifiers are applied on the same type as `#[repr(align(N), packed(M))]`,
+the alignment of the struct fields are decreased to be M. Then, the base alignment of the type is
+increased to be N.
+
+When a `#[repr(packed(M))]` struct transitively contains a field with `#[repr(align(N))]` type, depending on the
+target triplet, either:
+- The field is first `pad_to_align`. Then, the field is added to the struct with alignment decreased to M. The packing requirement overrides the alignment requirement. (GCC, `#[repr(Rust)]`, `#[repr(C)]` on gnu targets, `#[repr(system)]` on non-windows targets), or
+- The field is added to the struct with alignment increased to N. The alignment requirement overrides the packing requirement. (MSVC, `#[repr(C)]` on msvc targets, `#[repr(system)]` on windows targets)
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Historically the meaning of `#[repr(C)]` has been somewhat ambiguous. When someone puts `#[repr(C)]` on their struct, their intention could be one of three things:
```

Comment:
https://doc.rust-lang.org/reference/type-layout.html#the-c-representation actually documents the meaning of `repr(C)` quite clearly: it means types are laid out linearly, according to a fixed algorithm. So this RFC is proposing a breaking change, and unsafe code that relies on what is documented in the Reference might become subtly unsound if this RFC gets implemented.

That should at least be mentioned and discussed as a drawback.

Reply:
Isn't such a breaking change an explicit violation of Rust's stability policy? Though I can't actually find where it's documented ATM, my understanding is that breaking changes are only permitted in the following cases:
- The breakage is due to type inference
- The breakage is required in order to fix a security issue

Reply:
As described in https://github.com/rust-lang/rfcs/pull/3718#discussion_r1825075502, it should be simple to avoid this problem by introducing a new repr (in addition to the proposed `#[repr(system)]`) instead of changing the behavior of `#[repr(C)]`.

Reply:
It's definitely in a gray area. One cold also argue that the existing `repr(C)` is simply wrong/unsound, and we have to fix it to satisfy its promise of providing type layout compatible with the current target's C ABI.

Reply:
iirc someone suggested deprecating `repr(C)` entirely and replacing it with `repr(linear)`, `repr(really_C_bikeshed)`, and other `repr`s as necessary.